packages/devtools/lib/src/timeline/frame_flame_chart.dart

```diff
@@ -407,3 +430,142 @@ class FlameChartItem {
+      final timestamp = getTimestampForPosition(left + interval);
+      final gridItem = TimelineGridItem(left, interval, timestamp);
+
+      _gridItems.add(gridItem);
+      add(gridItem);
+
+      left += interval;
+    }
+  }
+
+  void _clear() {
+    clear();
+    _gridItems.clear();
+  }
+
+  num getTimestampForPosition(num gridItemEnd) {
+    return (gridItemEnd - flameChartInset) / flameChartWidth * _frameDurationMs;
+  }
+
+  void updateForZoom(num newZoomLevel, num newFlameChartWidth) {
+    if (zoomLevel == newZoomLevel) {
+      return;
+    }
+
+    flameChartWidth = newFlameChartWidth;
+    element.style.width = '${flameChartWidthWithInsets}px';
+
+    final log2ZoomLevel = log2(zoomLevel);
+    final log2NewZoomLevel = log2(newZoomLevel);
+
+    final gridZoomFactor = math.pow(2, log2NewZoomLevel);
+    final gridIntervalPx = baseGridInterval / gridZoomFactor;
+
+    /// The physical pixel width of the grid interval at [newZoomLevel].
+    final zoomedGridIntervalPx = gridIntervalPx * newZoomLevel;
+
+    if (log2NewZoomLevel == log2ZoomLevel) {
+      // Don't modify the first grid item. This item will have a fixed left of
+      // 0, width of [flameChartInset], and timestamp of '0.0 ms'.
+      for (int i = 1; i < _gridItems.length; i++) {
+        final currentItem = _gridItems[i];
+
+        final newLeft = flameChartInset + zoomedGridIntervalPx * (i - 1);
+        currentItem.update(newLeft, zoomedGridIntervalPx);
+      }
+    } else {
+      _clear();
+      _drawGrid(zoomedGridIntervalPx);
+    }
+
+    zoomLevel = newZoomLevel;
+  }
+}
+
+/// Describes a single item in the frame chart's timeline grid.
+///
+/// A single item consists of a line and a timestamp describing the location
+/// in the overall timeline [TimelineGrid].
+class TimelineGridItem extends CoreElement {
+  TimelineGridItem(this.currentLeft, this.currentWidth, this.timestamp)
+      : super('div', classes: 'flame-chart-grid-item') {
+    _initGridItem();
+  }
+
+  static const gridLineWidth = 1;
+  static const timestampPadding = 4;
+
+  num timestamp;
+  num currentLeft;
+  num currentWidth;
+
+  /// The timestamp label for this grid item.
+  CoreElement timestampLabel;
+
+  /// The line for this grid item.
+  CoreElement gridLine;
+
+  void _initGridItem() {
+    gridLine = div(c: 'grid-line');
+    add(gridLine);
+
+    timestampLabel = div(c: 'timestamp');
+    add(timestampLabel);
+
+    update(currentLeft, currentWidth);
+  }
+
+  void update(num left, num width) {
+    currentLeft = left;
+    currentWidth = width;
+
+    element.style
+      ..left = '${left}px'
+      ..width = '${width}px';
+
+    // Update [gridLine] position.
+    gridLine.element.style.left = '${width - gridLineWidth}px';
+
+    // Update [timestampLabel] position and text.
+    timestampLabel.element.style
+      ..left = '${timestampPadding}px'
+      ..width = '${width - 2 * timestampPadding}px';
+    timestampLabel.text =
```

Comment:
add a TODO to add more advanced logic for rounding the timestamps.

Reply:
Added todo and filed issue https://github.com/flutter/devtools/issues/329